src/chrome/internal/sources/sourceResolver.ts

```diff
@@ -0,0 +1,53 @@
+import { ILoadedSource } from './loadedSource';
+import { IUnresolvedSource, SourceToBeResolvedViaPath } from './unresolvedSource';
+import { newResourceIdentifierMap, IResourceIdentifier } from './resourceIdentifier';
+import { IComponent } from '../features/feature';
+import { ScriptParsedEvent } from '../../target/events';
+import { injectable, inject } from 'inversify';
+import { TYPES } from '../../dependencyInjection.ts/types';
+
+export interface EventsConsumedBySourceResolver {
+    onScriptParsed(listener: (scriptEvent: ScriptParsedEvent) => Promise<void>): void;
+}
+
+/**
+ * The SourceResolver listens to onScriptParsed events to build a map of paths to loaded sources. When an SourceToBeResolvedViaPath is created, it'll store a reference to this object,
+ * and use it when it tries to resolve the path to a loaded source
+ */
+
+ @injectable()
+export class SourceResolver implements IComponent {
+    private _pathToSource = newResourceIdentifierMap<ILoadedSource>();
+
+    public tryResolving<R>(sourceIdentifier: IResourceIdentifier,
```

Comment:
Why use this style with the callbacks, vs just throw in case of failure?

Reply:
I expect most callers of IUnresolvedSource.tryResolving which is what's calling SourceResolver.tryResolving to have to handle the failure case immediately instead of letting it bubble-up to the top level. By using this signature for the method, I'm forcing the consumers to have to supply that failure case (So they can't forget to do it). If I expected most consumers to want to let the failure bubble-up so the top-level can handle it, I'd throw an Exception instead.

BTW: No-one is using the default parameter of failedAction so I'm removing it...

Reply:
It seems like the caller should be the one to decide whether the failure case is handled or not. But not a big deal.